python/ray/data/datasource/parquet_datasource.py

```diff
@@ -254,6 +280,53 @@ def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
 
         return read_tasks
 
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the Parquet files encoding ratio.
+
+        To avoid OOMs, it is safer to return an over-estimate than an underestimate.
+        """
+        if not DatasetContext.get_current().decoding_size_estimation:
+            return PARQUET_ENCODING_RATIO_ESTIMATE_DEFAULT
+
+        # Sample a few rows from Parquet files to estimate the encoding ratio.
+        # Launch tasks to sample multiple files remotely in parallel.
+        # Evenly distributed to sample N rows in i-th row group in i-th file.
+        # TODO(ekl/cheng) take into account column pruning.
+        start_time = time.perf_counter()
+        num_files = len(self._pq_ds.pieces)
+        num_samples = int(num_files * PARQUET_ENCODING_RATIO_ESTIMATE_SAMPLING_RATIO)
+        min_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MIN_NUM_SAMPLES, num_files
+        )
+        max_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MAX_NUM_SAMPLES, num_files
+        )
+        num_samples = max(min(num_samples, max_num_samples), min_num_samples)
+
+        # Evenly distributed to choose which file to sample, to avoid biased prediction
+        # if data is skewed.
+        file_samples = [
+            self._pq_ds.pieces[idx]
+            for idx in np.linspace(0, num_files - 1, num_samples).astype(int).tolist()
+        ]
+
+        sample_piece = cached_remote_fn(_sample_piece)
+        futures = []
+        for idx, sample in enumerate(file_samples):
+            # Sample i-th row group in i-th file.
```

Comment:
I assume that this is to avoid row group bias within Parquet files; would it be better to just randomly sample a row group within each Parquet file, or are we wanting to get ~complete coverage of all row groups across Parquet files (assuming they're ~uniform)? If the latter is the goal, that sounds good.

Reply:
One goal here is determinism, so that the size_bytes() doesn't seem to randomly change each time the data is read.

Reply:
Yep that makes sense, sounds good. 👍

Reply:
Yeah I think determinism is better here, some prevous comments - https://github.com/ray-project/ray/pull/26623#discussion_r924086711 .